[FEATURE-13]: support relators role notes as part of creators text

davorpa opened the pull request:
### What this PR do?

Add feature(s)

### Description

Recently we introduced the `ABBR.:` note in [markdown sources](https://github.com/EbookFoundation/free-programming-books/search?q=trl.%3A+is%3Apr&type=issues) to support relator abbreviations attached to each creator.

e.g. for creators of type translator: 
```markdown
`trl.:` an awesome translator
```
More relators terms: https://loc.gov/marc/relators/relaterm.html

This PR adds supports to them

### Context

I attach two files in a zip: [fpb-parser-data.zip](https://github.com/EbookFoundation/free-programming-books-parser/files/9623949/fpb-parser-data.zip). Compares them to test differences. It seems all right

davorpa committed:
feat: support text role notes for creators

e.g. for creators of type translator: `trl.:` ABC

davorpa commented on index.js:
```diff
@@ -121,23 +155,10 @@ function parseListItem(listItem) {
         });
         // entry.otherLinks = [...entry.otherLinks, {title: i.children[0].value, url: i.url}];      // <-- i wish i could get this syntax to work with arrays
-      }
-      if (i.type === "text" && i.value.indexOf("(") !== -1) {
-        // notes found (currently assumes no nested parentheses)
-        if (entry.notes === undefined) entry.notes = [];
-        leftParen = i.value.indexOf("(");
-        while (leftParen != -1) {
-          rightParen = i.value.indexOf(")", leftParen);
-          if (rightParen === -1) {
-            // there must be some *emphasis* found
-            s += i.value.slice(leftParen);
-            break;
-          }
-          entry.notes.push(i.value.slice(leftParen + 1, rightParen));
-          leftParen = i.value.indexOf("(", rightParen);
-        }
```
moved to lines 120-134

https://github.com/davorpa/free-programming-books-parser/blob/9ea1a0307338552583c6bb712202f6540de9135e/index.js#L120-L134

davorpa commented on index.js:
```diff
@@ -93,25 +93,59 @@ function parseListItem(listItem) {
-      if (i.type === "text" && i.value.slice(0, 3) === " - ") {
-        // author found
-        let parenIndex = i.value.indexOf("(");
-        if (parenIndex === -1) {
-          entry.author = i.value.slice(3).trim();
-        } else {
-          entry.author = i.value.slice(3, parenIndex).trim(); // go from " - " until the first "("
```
`trim()` moved to lines 197-198

https://github.com/davorpa/free-programming-books-parser/blob/9ea1a0307338552583c6bb712202f6540de9135e/index.js#L197-L198

eshellman commented on index.js:
```diff
@@ -93,25 +93,59 @@ function parseListItem(listItem) {
       }
+      if (insideAuthors && i.type === "inlineCode") {
+        // author role found. append rebuilding markdown format and then move on
+        entry.author += "`" + i.value + "`";
```
seems like the values accepted as relators should be limited!

davorpa replied:
Are you talking about limit to `trl.:` ? I think this is more to take into account reviewing fpb PRs or in the fpb-linter

eshellman replied:
for the parser, mostly don't want it to be tripped up by syntax we haven't anticipated.

davorpa replied:
I have not fully understood what you meant by the last two posts. Could you develop the premise better or tell me how I should code it?

davorpa commented:
Force-push (https://github.com/EbookFoundation/free-programming-books-parser/pull/13#event-7454559181) rebase was for integrate #8 functions

eshellman commented:
Suppose someone enters
Charles Windsor, `a.k.a.:` King Charles
or
Megan `Thee` Stallion
or
Main Author. See also [`wierd site`](http://link)

davorpa commented:
> Suppose someone enters Charles Windsor, `a.k.a.:` King Charles or Megan `Thee` Stallion or Main Author. See also [`wierd site`](http://link)

So.... if I understand well... relator terms should be the first token after a comma `,` or if it is the first, authors field should be empty? If not, a flag, `e.g.` `manualReviewRequired = true` like happens with links needs be added to the record?

![image](https://user-images.githubusercontent.com/3125580/193246722-dc02db2f-94e8-434c-ae41-cc36b52b7b9b.png)
![image](https://user-images.githubusercontent.com/3125580/193246884-0060e3f2-9bdb-4b84-ad0e-77a012dc812f.png)

Moreover... the unique offending case could be the 2nd because...
- The 1st is a valid creator slug if we dont take into account value validity
- The 3th case is already comtemplated since it's a link node with inline code inside and therefore this is a note.

eshellman commented:
My point is that verifying validity of a relator token is easy.

davorpa commented:
> My point is that verifying validity of a relator token is easy.

Then, I will also put a list like `languages.js` defining all of them. Thanks for the suggestions

davorpa committed:
feat: introduce `hasAuthorWarnings` flag

This flag validates the relator term and it position inside each creator string

davorpa committed:
feat: add marc relator api to limit valid creator roles

davorpa commented:
> My point is that verifying validity of a relator token is easy.

Done
